File: webrtc-jni/src/main/cpp/src/api/RTCRtpEncodingParameters.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "api/RTCRtpEncodingParameters.h"
#include "JavaClasses.h"
#include "JavaEnums.h"
#include "JavaObject.h"
#include "JavaPrimitive.h"
#include "JavaRef.h"
#include "JNI_WebRTC.h"

namespace jni
{
	namespace RTCRtpEncodingParameters
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const webrtc::RtpEncodingParameters & parameters)
		{
			const auto javaClass = JavaClasses::get<JavaRTCRtpEncodingParametersClass>(env);

			jobject object = env->NewObject(javaClass->cls, javaClass->ctor);
			env->SetObjectField(object, javaClass->active, Boolean::create(env, parameters.active));

			if (parameters.ssrc.has_value()) {
				env->SetObjectField(object, javaClass->ssrc, Long::create(env, parameters.ssrc.value()));
			}
			if (parameters.min_bitrate_bps.has_value()) {
				env->SetObjectField(object, javaClass->minBitrate, Integer::create(env, parameters.min_bitrate_bps.value()));
			}
			if (parameters.max_bitrate_bps.has_value()) {
				env->SetObjectField(object, javaClass->maxBitrate, Integer::create(env, parameters.max_bitrate_bps.value()));
			}
			if (parameters.max_framerate.has_value()) {
				env->SetObjectField(object, javaClass->maxFramerate, Double::create(env, parameters.max_framerate.value()));
			}
			if (parameters.scale_resolution_down_by.has_value()) {
				env->SetObjectField(object, javaClass->scaleResolution, Double::create(env, parameters.scale_resolution_down_by.value()));
			}

			return JavaLocalRef<jobject>(env, object);
		}
		
		webrtc::RtpEncodingParameters toNative(JNIEnv * env, const JavaRef<jobject> & parameters)
		{
			const auto javaClass = JavaClasses::get<JavaRTCRtpEncodingParametersClass>(env);

			JavaObject obj(env, parameters);

			auto active = obj.getObject(javaClass->active);
			auto ssrc = obj.getObject(javaClass->ssrc);
			auto minBitrate = obj.getObject(javaClass->minBitrate);
			auto maxBitrate = obj.getObject(javaClass->maxBitrate);
			auto maxFramerate = obj.getObject(javaClass->maxFramerate);
			auto scaleResolution = obj.getObject(javaClass->scaleResolution);

			webrtc::RtpEncodingParameters params;

			if (active.get()) {
				params.active = Boolean::getValue(env, active);
			}
			if (ssrc.get()) {
				params.ssrc = static_cast<uint32_t>(Long::getValue(env, ssrc));
			}
			if (minBitrate.get()) {
				params.min_bitrate_bps.emplace(Integer::getValue(env, minBitrate));
			}
			if (maxBitrate.get()) {
				params.max_bitrate_bps.emplace(Integer::getValue(env, maxBitrate));
			}
			if (maxFramerate.get()) {
				params.max_framerate.emplace(Double::getValue(env, maxFramerate));
			}
			if (scaleResolution.get()) {
				params.scale_resolution_down_by.emplace(Double::getValue(env, scaleResolution));
			}

			return params;
		}

		JavaRTCRtpEncodingParametersClass::JavaRTCRtpEncodingParametersClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG"RTCRtpEncodingParameters");

			ctor = GetMethod(env, cls, "<init>", "()V");
			
			ssrc = GetFieldID(env, cls, "ssrc", LONG_SIG);
			active = GetFieldID(env, cls, "active", BOOLEAN_SIG);
			minBitrate = GetFieldID(env, cls, "minBitrate", INTEGER_SIG);
			maxBitrate = GetFieldID(env, cls, "maxBitrate", INTEGER_SIG);
			maxFramerate = GetFieldID(env, cls, "maxFramerate", DOUBLE_SIG);
			scaleResolution = GetFieldID(env, cls, "scaleResolutionDownBy", DOUBLE_SIG);
		}
	}
}

File: webrtc/src/main/java/dev/onvoid/webrtc/RTCRtpEncodingParameters.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc;

/**
 * Describes encoding options of an {@link RTCRtpSender}.
 *
 * @author Alex Andres
 */
public class RTCRtpEncodingParameters {

	/**
	 * If unset, a value is chosen by the implementation.
	 * <br>
	 * Note that the chosen value is NOT returned by GetParameters, because it
	 * may change due to an SSRC conflict, in which case the conflict is handled
	 * internally without any event. Another way of looking at this is that an
	 * unset SSRC acts as a "wildcard" SSRC.
	 */
	public Long ssrc;

	/**
	 * Indicates that this encoding is actively being sent. Setting it to false
	 * causes this encoding to no longer be sent. Setting it to true causes this
	 * encoding to be sent.
	 */
	public Boolean active;

	/**
	 * When present, indicates the maximum bitrate that can be used to send this
	 * encoding. If unset, there is no maximum bitrate.
	 */
	public Integer maxBitrate;

	/**
	 * When present, indicates the minimum bitrate that can be used to send this
	 * encoding. If unset, there is no minimum bitrate.
	 */
	public Integer minBitrate;

	/**
	 * When present, indicates the maximum frame rate that can be used to send
	 * this encoding, in frames per second.
	 */
	public Double maxFramerate;

	/**
	 * Only present if the sender's kind is "video". The video's resolution will
	 * be scaled down in each dimension by the given value before sending. For
	 * example, if the value is 2.0, the video will be scaled down by a factor
	 * of 2 in each dimension, resulting in sending a video of one quarter the
	 * size. If the value is 1.0, the video will not be affected. The value must
	 * be greater than or equal to 1.0. By default, the sender will not apply
	 * any scaling
	 */
	public Double scaleResolutionDownBy;


	/**
	 * Creates an instance of RTCRtpEncodingParameters.
	 */
	public RTCRtpEncodingParameters() {
		active = true;
	}

	@Override
	public String toString() {
		return "RTCRtpEncodingParameters{" + "ssrc=" + ssrc + ", active="
				+ active + ", maxBitrate=" + maxBitrate + ", minBitrate="
				+ minBitrate + ", maxFramerate=" + maxFramerate
				+ ", scaleResolutionDownBy=" + scaleResolutionDownBy + '}';
	}
}
